staging/src/k8s.io/client-go/tools/portforward/portforward.go

```diff
@@ -231,6 +233,8 @@ func (pf *PortForwarder) forward() error {
 	case <-pf.stopChan:
 	case <-pf.streamConn.CloseChan():
 		runtime.HandleError(errors.New("lost connection to pod"))
+	case err := <-pf.errorChan:
```

Comment:
1. in the first case, a stop was called.  portforward should probably shut down its ports on this for the library (pr wanted), but the intent is likely "process shutting down" when we wrote it.
2. in the second case, the streamconnection was lost
3. in this new case, the local handleConnection will stop https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L356, but the callsite will restart it: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L302-L312

if I'm reading this code correctly, this would return an error back and start listening on the port again.  the tests are for the handleConnection, but miss the waitForConnection which will restart it indefinitely.

Reply:
This error actually gets returned all the way up to the kubectl command itself and stops execution.

[Return from `forward()`](https://github.com/kubernetes/kubernetes/blob/bd06c694000cb865fb1e94cc2857e84dcbb43d74/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L237)

[Return from `ForwardPorts`](https://github.com/kubernetes/kubernetes/blob/bd06c694000cb865fb1e94cc2857e84dcbb43d74/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L200)

[Return from `ForwardPorts`](https://github.com/kubernetes/kubernetes/blob/bd06c694000cb865fb1e94cc2857e84dcbb43d74/staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go#L143)

[Return from `RunPortForward`](https://github.com/kubernetes/kubernetes/blob/bd06c694000cb865fb1e94cc2857e84dcbb43d74/staging/src/k8s.io/kubectl/pkg/cmd/portforward/portforward.go#L415)

Reply:
I  think that still races for whether the ForwardPort go route runs before the reconnect from waitForConnection.

I'm supportive of trying to return errors.  I just want deterministic behavior when we make this change.  Closing a stop channel (or even better a context) that is honored throughout the call chain is an approach that will make this deterministic and non-leaky.